agent/configuration-build-derive/src/derive.rs

```diff
@@ -0,0 +1,433 @@
+    let setters = data.fields.iter().filter_map(|field| {
+        let doc = field.attrs.iter().filter(|v| {
+            v.parse_meta()
+                .map(|meta| meta.path().is_ident("doc"))
+                .unwrap_or(false)
+        });
+        let field_name = match field.ident.as_ref() {
+            Some(ident) => ident.to_string(),
+            None => return None,
+        };
+        let field_ident = Ident::new(&field_name, Span::call_site());
+        let template_field_name = format!("{}{}", field_name, "_template");
+        let template_ident = Ident::new(&template_field_name, Span::call_site());
+        let ty = field.ty.clone();
+        Some(quote! {
+            #(#doc)*
+            #[inline(always)]
+            pub fn #field_ident<T>(&mut self, #field_ident: T) -> &mut Self
+            where
+            T: Into<#ty>{
+                self.#field_ident = model::ConfigValue::Value(#field_ident.into());
+                self
+            }
+
+            #[inline(always)]
+            pub fn #template_ident<S1,S2>(&mut self, resource: S1, field: S2) -> &mut Self
+            where
+            S1: Into<String>,
+            S2: Into<String>,
+            {
+                self.#field_ident = model::ConfigValue::TemplatedString(format!("${{{}.{}}}", resource.into(), field.into()));
+                self
+            }
+        })
+    });
+    let fns = quote! {
+        #(#setters)*
+    };
+
+    // Add the build function to the builders.
+    let build = match crd_type.as_str() {
+        "Test" => {
+            quote! {
+                #[derive(Debug, Default, serde::Serialize)]
+                pub struct #build_ident{
+                    #(#fields)*
+                    depends_on: Vec<String>,
+                    resources: Vec<String>,
+                    labels: std::collections::BTreeMap<String,String>,
+                    image: Option<String>,
+                    image_pull_secret: Option<String>,
+                    secrets: std::collections::BTreeMap<String, model::SecretName>,
+                    retries: Option<u32>,
+                    keep_running: Option<bool>,
+                    capabilities: Vec<String>,
+                }
+
+                impl #build_ident{
+                    pub fn new() -> #build_ident {
+                        Default::default()
+                    }
+
+                    #fns
+
+                    pub fn depends_on<S1>(&mut self, depends_on: S1) -> &mut Self
+                    where
+                    S1: Into<String> {
+                        self.depends_on.push(depends_on.into());
+                        self
+                    }
+
+                    pub fn set_depends_on(&mut self, depends_on: Option<Vec<String>>) -> &mut Self {
+                        self.depends_on = depends_on.unwrap_or_default();
+                        self
+                    }
```

Comment:
These `set_...` all look almost the same to me. Why do we take options for the argument and not just the type itself with `fn set_<field>` and we take the type itself and wrap it in an `Some` for `fn <field>`?

I wonder if there's a way to use `proc_macro` to generate and append these `set_*` setters by iterating over all the fields in the struct. So maybe something like (excuse the ugly formatting):
```
fn gen_set(crd: DataStruct) -> TokenStream {
    let mut set_stream = TokenStream2::default();
    for field in crd.fields {
        let fname=field.name;
        let ftype=field.type;
        set_stream.extend(
           quote! { 
                 fn set_#fname<S1>(&mut self, #fname: #ftype) -> &mut Self
                 where .... { 
                    self.#fname = #fname;
                    self
                 }
           }
       ) 
    }
    quote!{
        #set_stream
    }
}

...
impl #build_ident{
    #gen_set(#build_ident)
}
```

That way we wouldn't need to edit this whenever add a new optional field to the test or resource CRD.